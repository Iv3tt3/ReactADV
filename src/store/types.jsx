export const AUTH_LOGIN_PENDING = "auth/login/pending";
export const AUTH_LOGIN_FULFILLED = "auth/login/fulfilled";
export const AUTH_LOGIN_REJECTED = "auth/login/rejected";

export const AUTH_LOGOUT = "auth/logout";

export const UI_RESET_ERROR = "ui/reset-error";

export const ADVERTS_LOADED_PENDING = "adverts/loaded/pending";
export const ADVERTS_LOADED_FULFILLED = "adverts/loaded/fulfilled";
export const ADVERTS_LOADED_REJECTED = "adverts/loaded/rejected";

export const ADVERTS_CREATED_PENDING = "adverts/created/pending";
export const ADVERTS_CREATED_FULFILLED = "adverts/created/fulfilled";
export const ADVERTS_CREATED_REJECTED = "adverts/created/rejected";

export const ADVERT_DETAIL_PENDING = "advert/detail/pending";
export const ADVERT_DETAIL_FULFILLED = "advert/detail/fulfilled";
export const ADVERT_DETAIL_REJECTED = "advert/detail/rejected";

export const TAGS_LOADED_PENDING = "tags/loaded/pending";
export const TAGS_LOADED_FULFILLED = "tags/loaded/fulfilled";
export const TAGS_LOADED_REJECTED = "tags/loaded/rejected";

export const TAGS_SELECTED = "tags/selected";
export const TAGS_SELECTED_ADD = "tags/selected/add";