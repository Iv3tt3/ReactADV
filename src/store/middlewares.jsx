// eslint-disable-next-line no-unused-vars
export const errorMiddleware = (router, redirectMap) => (store) => (next) => (action) => {
  const result = next(action);

  if (!action.error) {
    return result;
  }
  const redirect = redirectMap[action.payload.status]
  if (redirect) {
    router.navigate(redirect);
  }
  return result;
};
