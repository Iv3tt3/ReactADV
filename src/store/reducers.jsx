import * as t from "./types";

export const defaultState = {
  auth: false,
  ui: {
    pending: false,
    error: null,
  },
  adverts: {
    loaded: false, 
    data: [],
  },
};

export function auth(state = defaultState.auth, action) {
  switch (action.type) {
    case t.AUTH_LOGIN_FULFILLED:
      return true;
    case t.AUTH_LOGOUT:
      return false;
    default:
      return state;
  }
}

export function adverts(state = defaultState.adverts, action) {
  switch (action.type) {
    case t.ADVERTS_LOADED_FULFILLED:
      return {...state, loaded: true, data: action.payload};
    case t.ADVERTS_CREATED_FULFILLED:
      return {...state, data: [...state.data, action.payload]};
    case t.ADVERT_DETAIL_FULFILLED:
      return {...state, data: [action.payload]}
    default:
      return state;
  }
}

export function ui(state = defaultState.ui, action) {
  if (action.error) {
    return { ...state, pending: false, error: action.payload };
  }

  if (action.type === t.UI_RESET_ERROR) {
    return { ...state, error: null };
  }

  if (action.type.endsWith('/pending')) {
    return { ...state, pending: true };
  }

  if (action.type.endsWith('/fulfilled')) {
    return { ...state, pending: false, error: null };
  }

  return state;
}