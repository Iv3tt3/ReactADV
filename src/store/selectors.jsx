export const getIsLogged = state => state.auth;

export const getAdverts = state => state.adverts;

export const getAdvert = advertId => state => 
    getAdverts(state).find(advert => advert.id === advertId);