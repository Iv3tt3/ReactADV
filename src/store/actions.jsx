import { areAdvertsLoaded, getAdvert, getTagsFromAds } from "./selectors.jsx";
import * as t from "./types.jsx";

export const authLoginPending = () => ({
  type: t.AUTH_LOGIN_PENDING,
});

export const authLoginFulfilled = () => ({
  type: t.AUTH_LOGIN_FULFILLED,
});

export const authLoginRejected = (error) => ({
  type: t.AUTH_LOGIN_REJECTED,
  payload: error,
  error: true,
});

export const authLogin = (credentials, checked) => {
  return async function (dispatch, _getState, { services: { auth }, router }) {
    try {
      dispatch(authLoginPending());
      await auth.login(credentials, checked);
      dispatch(authLoginFulfilled());
      router.navigate(router.state.location.state?.pathname || "/");
    } catch (error) {
      dispatch(authLoginRejected(error));
    }
  };
};

export const authLogout = () => ({
  type: t.AUTH_LOGOUT,
});

export const logoutAction = () => {
  return function (dispatch, getState, { services: { auth } }) {
    auth.logout();
    dispatch(authLogout());
  };
};

export const uiResetError = () => ({
  type: t.UI_RESET_ERROR,
});

export const advertsLoadedPending = () => ({
  type: t.ADVERTS_LOADED_PENDING,
});

export const advertsLoadedFulfilled = (adverts) => ({
  type: t.ADVERTS_LOADED_FULFILLED,
  payload: adverts,
});

export const advertsLoadedRejected = (error) => ({
  type: t.ADVERTS_LOADED_REJECTED,
  payload: error,
  error: true,
});

export const loadAdverts = () => {
  return async function (dispatch, getState, { services: { adverts } }) {
    try {
      const state = getState();
      if (areAdvertsLoaded(state)) {
        return;
      }
      dispatch(advertsLoadedPending());
      const advertsData = await adverts.getAdverts();
      dispatch(advertsLoadedFulfilled(advertsData));
      dispatch(loadTags());
    } catch (error) {
      dispatch(advertsLoadedRejected(error));
    }
  };
};

export const advertCreatedPending = () => ({
  type: t.ADVERTS_CREATED_PENDING,
});

export const advertCreatedFulfilled = (advert) => ({
  type: t.ADVERTS_CREATED_FULFILLED,
  payload: advert,
});

export const advertCreatedRejected = (error) => ({
  type: t.ADVERTS_CREATED_REJECTED,
  payload: error,
  error: true,
});

export const createAd = (advert) => {
  return async function (
    dispatch,
    getState,
    { services: { adverts }, router }
  ) {
    try {
      dispatch(advertCreatedPending());
      const { id } = await adverts.postAdvert(advert);
      const advertInfo = await adverts.getAdvert(id);
      dispatch(advertCreatedFulfilled(advertInfo));
      router.navigate("/");
    } catch (error) {
      dispatch(advertCreatedRejected(error));
    }
  };
};

export const advertDetailPending = () => ({
  type: t.ADVERT_DETAIL_PENDING,
});

export const advertDetailFulfilled = (advert) => ({
  type: t.ADVERT_DETAIL_FULFILLED,
  payload: advert,
});

export const advertDetailRejected = (error) => ({
  type: t.ADVERT_DETAIL_REJECTED,
  payload: error,
  error: true,
});

export const loadAdvert = (advertId) => {
  return async function (dispatch, getState, { services: { adverts } }) {
    try {
      const state = getState();
      if (getAdvert(advertId)(state)) {
        return;
      }
      dispatch(advertDetailPending());
      const advert = await adverts.getAdvert(advertId);
      dispatch(advertDetailFulfilled(advert));
    } catch (error) {
      dispatch(advertDetailRejected(error));
    }
  };
};

export const advertDelete = (adverts) => ({
  type: t.ADVERT_DELETE,
  payload: adverts,
});

export const advertDeleteRejected = (error) => ({
  type: t.ADVERT_DELETE_REJECTED,
  payload: error,
  error: true,
});

export const deleteAdvert = (advertId) => {
  return async function (
    dispatch,
    getState,
    { services: { adverts }, router }
  ) {
    try {
      const state = getState();
      await adverts.deleteAdvert(advertId);
      const filterAdverts = state.adverts.data.filter(
        (advert) => advert.id !== advertId
      );
      dispatch(advertDelete(filterAdverts));
      dispatch(loadTags());
      router.navigate("/");
    } catch (error) {
      dispatch(advertDeleteRejected(error));
    }
  };
};

export const tagsLoadedPending = () => ({
  type: t.TAGS_LOADED_PENDING,
});

export const tagsLoadedFulfilled = (tags) => ({
  type: t.TAGS_LOADED_FULFILLED,
  payload: tags,
});

export const tagsLoadedRejected = (error) => ({
  type: t.TAGS_LOADED_REJECTED,
  payload: error,
  error: true,
});

export const loadTags = () => {
  return async function (dispatch, getState) {
    try {
      const state = getState();
      //To resolve problem with API. Api returns 4 tags when there is 6 differen tags in my ads. I will use info in redux state and avoid API request

      // if (areTagsLoaded(state)) {
      //   return;
      // }
      //const tagsData = await adverts.getTags();
      dispatch(tagsLoadedPending());
      const tagsData = getTagsFromAds(state);
      dispatch(tagsLoadedFulfilled(tagsData));
    } catch (error) {
      dispatch(tagsLoadedRejected(error));
    }
  };
};

export const selectedTags = (tags) => ({
  type: t.TAGS_SELECTED,
  payload: tags,
});

export const addSelectedTag = (tag) => ({
  type: t.TAGS_SELECTED_ADD,
  payload: tag,
});

export const addTag = (tag) => ({
  type: t.TAGS_ADD,
  payload: tag,
});
