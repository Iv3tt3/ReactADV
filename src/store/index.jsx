import { createStore, combineReducers, applyMiddleware } from "redux";
import { composeWithDevTools } from '@redux-devtools/extension';
import { withExtraArgument } from "redux-thunk";

import * as reducers from "./reducers";
import * as actionCreators from "./actions";
import * as auth from "../pages/auth/service";
import * as adverts from "../pages/adverts/service";


const reducer = combineReducers(reducers);

const composeEnhancers = composeWithDevTools({ actionCreators });

export default function configureStore(preloadedState) {
  const store = createStore(
    reducer,
    preloadedState,
    composeEnhancers(applyMiddleware(withExtraArgument({ services: { auth, adverts } })))
  );
  return store;
}
