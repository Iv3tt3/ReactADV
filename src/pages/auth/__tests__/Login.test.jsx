import { fireEvent, render, screen } from "@testing-library/react";
import { describe, expect, test, vi } from "vitest";
import LoginPage from "../Login";
import { Provider } from "react-redux";
import { MemoryRouter } from "react-router-dom";
import { authLogin } from "../../../store/actions";

vi.mock("../../../store/actions");

describe("LoginPage", () => {
  const state = { ui: { pending: false, error: null } };
  const store = {
    dispatch: () => {},
    getState: () => state,
    subscribe: () => {},
  };
  const renderComponent = () =>
    render(
      <Provider store={store}>
        <MemoryRouter>
          <LoginPage />
        </MemoryRouter>
      </Provider>
    );

  test("snapshot", () => {
    const { container } = renderComponent();
    expect(container).toMatchSnapshot();
  });

  test("should dispatch authLogin action"),
    () => {
      const email = "email";
      const password = "password";
      const checked = "checked";

      const emailInput = screen.getByRole("textbox", {
        name: /email/i,
      });
      const passwordInput = screen.getByText(/password/i);
      const checkInput = screen.getByRole("checkbox", {
        name: /stay logged in/i,
      });
      const submitButton = screen.getByRole("button", {
        name: /log in/i,
      });
      expect(submitButton).toBeDisabled();

      fireEvent.change(emailInput, { target: { value: email } });
      fireEvent.change(passwordInput, { target: { value: password } });
      fireEvent.click(checkInput, { target: { value: checked } });

      expect(submitButton).toBeEnabled();

      fireEvent.click(submitButton);

      expect(authLogin).toHaveBeenCalledWith({email, password}, checked)

    };
});
