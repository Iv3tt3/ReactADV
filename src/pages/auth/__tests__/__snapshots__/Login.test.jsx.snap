// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`LoginPage > snapshot 1`] = `
<div>
  <div>
    <div
      class="_layout_a1a5c1"
    >
      <header>
        <a
          href="/"
        >
          <svg
            viewBox="0 0 600 150"
            width="300"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M 32.6 55.3 L 20.2 2.7 L 41.3 2.7 L 58 75.5 L 37.3 75.5 L 36.2 70.8 L 26.9 70.8 L 29.6 55.3 L 32.6 55.3 Z M 137.3 17.6 L 117.8 17.6 L 112.3 3.5 L 141.1 3.5 A 54.574 54.574 0 0 1 147.291 3.83 Q 150.401 4.186 153.025 4.925 A 26.44 26.44 0 0 1 156 5.95 Q 161.9 8.4 164.8 14.2 A 23.652 23.652 0 0 1 166.567 19.141 Q 167.187 21.661 167.467 24.673 A 57.519 57.519 0 0 1 167.7 30 Q 167.7 38.118 165.422 43.578 A 18.092 18.092 0 0 1 161.4 49.65 A 19.965 19.965 0 0 1 154.117 54.082 Q 149.104 55.893 142.2 56 L 142.2 40.8 Q 144.5 39.7 145.65 37.05 Q 146.432 35.249 146.682 32.547 A 29.863 29.863 0 0 0 146.8 29.8 A 49.386 49.386 0 0 0 146.733 27.129 Q 146.561 23.969 145.95 22.1 A 7.266 7.266 0 0 0 145.194 20.46 A 4.993 4.993 0 0 0 143.05 18.55 A 8.503 8.503 0 0 0 141.446 18.009 Q 139.728 17.6 137.3 17.6 Z M 90.3 67.6 L 90.3 2.7 L 108.3 0 L 108.3 75.5 L 93 75.5 L 90.3 67.6 Z M 241.9 18 L 244.6 25.8 L 244.6 87.9 L 226.6 90.5 L 226.6 18 L 241.9 18 Z M 25.9 48.4 L 20.9 75.5 L 0 75.5 L 16.6 9.4 L 25.9 48.4 Z M 250.1 31.2 L 249.6 31.2 L 249.6 18.3 Q 252.7 17.2 256 17.2 Q 260.694 17.2 263.968 18.912 A 12.201 12.201 0 0 1 265.05 19.55 A 16.07 16.07 0 0 1 270.252 25.246 A 19.261 19.261 0 0 1 270.65 26 A 30.331 30.331 0 0 1 272.863 31.983 A 37.23 37.23 0 0 1 273.55 35.25 Q 274.4 40.4 274.4 46.1 A 83.338 83.338 0 0 1 274.04 54.149 Q 273.259 62.166 270.782 67.057 A 17.775 17.775 0 0 1 269.3 69.5 A 16.571 16.571 0 0 1 257.348 76.329 A 24.608 24.608 0 0 1 254.4 76.5 Q 253.3 76.5 252.1 76.35 Q 250.9 76.2 249.6 76 L 249.6 62.5 Q 252.871 62.091 254.202 59.209 A 8.471 8.471 0 0 0 254.7 57.8 Q 255.683 54.048 255.787 47.265 A 107.725 107.725 0 0 0 255.8 45.6 Q 255.8 39.737 255.139 36.435 A 13.348 13.348 0 0 0 254.6 34.45 A 5.894 5.894 0 0 0 253.717 32.825 Q 252.574 31.378 250.598 31.219 A 6.235 6.235 0 0 0 250.1 31.2 Z M 85.2 17.7 L 85.2 31.3 Q 81.3 31.7 80.2 35.9 Q 79.217 39.652 79.113 46.435 A 107.725 107.725 0 0 0 79.1 48.1 Q 79.1 53.963 79.761 57.265 A 13.348 13.348 0 0 0 80.3 59.25 A 5.894 5.894 0 0 0 81.183 60.875 Q 82.326 62.322 84.302 62.481 A 6.235 6.235 0 0 0 84.8 62.5 L 85.2 62.5 L 85.2 75.4 Q 82.2 76.5 78.9 76.5 Q 74.206 76.5 70.932 74.788 A 12.201 12.201 0 0 1 69.85 74.15 A 16.07 16.07 0 0 1 64.648 68.454 A 19.261 19.261 0 0 1 64.25 67.7 A 30.331 30.331 0 0 1 62.037 61.717 A 37.23 37.23 0 0 1 61.35 58.45 Q 60.5 53.3 60.5 47.6 A 82.535 82.535 0 0 1 60.863 39.551 Q 61.652 31.534 64.153 26.643 A 17.743 17.743 0 0 1 65.65 24.2 A 16.711 16.711 0 0 1 77.862 17.338 A 24.422 24.422 0 0 1 80.5 17.2 A 17.91 17.91 0 0 1 82.206 17.284 A 20.672 20.672 0 0 1 82.8 17.35 Q 84 17.5 85.2 17.7 Z M 198.4 76.4 L 198.4 61.7 Q 201.673 59.518 201.97 49.068 A 76.381 76.381 0 0 0 202 46.9 A 70.326 70.326 0 0 0 201.873 42.507 Q 201.325 33.787 198.4 32 L 198.4 17.3 A 33.709 33.709 0 0 1 204.839 18.147 Q 208.422 19.005 211.115 20.694 A 15.78 15.78 0 0 1 215.5 24.75 A 21.361 21.361 0 0 1 218.404 30.491 Q 220.3 36.016 220.559 44.156 A 83.1 83.1 0 0 1 220.6 46.8 A 71.021 71.021 0 0 1 220.208 54.549 Q 219.328 62.536 216.474 67.37 A 17.771 17.771 0 0 1 215.5 68.85 Q 211.124 74.814 201.667 76.102 A 38.443 38.443 0 0 1 198.4 76.4 Z M 193.4 61.8 L 193.4 76.4 Q 185.576 76.009 180.685 72.918 A 15.944 15.944 0 0 1 176.3 68.85 A 21.361 21.361 0 0 1 173.396 63.109 Q 171.5 57.584 171.241 49.444 A 83.1 83.1 0 0 1 171.2 46.8 A 70.914 70.914 0 0 1 171.592 39.055 Q 172.472 31.078 175.324 26.273 A 17.534 17.534 0 0 1 176.3 24.8 Q 180.663 18.897 190.076 17.605 A 38.647 38.647 0 0 1 193.4 17.3 L 193.4 31.9 A 4.687 4.687 0 0 0 192.053 33.237 Q 191.217 34.453 190.715 36.433 A 16.844 16.844 0 0 0 190.65 36.7 A 25.074 25.074 0 0 0 190.202 39.239 Q 189.857 41.92 189.808 45.651 A 95.873 95.873 0 0 0 189.8 46.9 A 70.326 70.326 0 0 0 189.927 51.293 Q 190.475 60.013 193.4 61.8 Z M 117.8 75.5 L 117.8 22.6 L 137.3 22.6 L 137.3 75.5 L 117.8 75.5 Z"
              fill="#558c62"
              fill-rule="evenodd"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-width="0.4mm"
              vector-effect="non-scaling-stroke"
            />
          </svg>
        </a>
        <nav />
      </header>
      <main>
        <h1>
          Login to your account
        </h1>
        <form
          class="_loginForm_fd6bf1"
        >
          <div
            class="_textInput_fd6bf1"
          >
            <label>
              <span>
                Email
              </span>
              <input
                autocomplete="off"
                class="formField-input"
                id="email"
                name="email"
                placeholder="admin@example.com"
                type="text"
                value=""
              />
            </label>
          </div>
          <div
            class="_textInput_fd6bf1"
          >
            <label>
              <span>
                Password
              </span>
              <input
                autocomplete="off"
                class="formField-input"
                id="password"
                name="password"
                placeholder="123456"
                type="password"
                value=""
              />
            </label>
          </div>
          <label>
            Stay Logged in
            <input
              class="_checkInput_fd6bf1"
              id="checkbox"
              type="checkbox"
            />
          </label>
          <button
            class="sc-aYaIB jpekxv"
            disabled=""
            type="submit"
          >
            Log in
          </button>
        </form>
      </main>
      <footer>
        @2024 Adpop Iv3tt3
      </footer>
    </div>
  </div>
</div>
`;
